### PR TITLE
chore!: Restrict packages to contain at most a single contract

### LINKED ISSUE #2329: Enforce a single Noir contract per package
### Problem

We currently support an unlimited number of contracts within a single Noir package. We've decided that we'd like there to be a single contract per package instead.

### Happy Case

Multiple contracts should result in a parsing error. We can then replace the functions to pull contracts from the defmap to instead return a single contract.

As there's only a single contract we can also remove the name field from `contract <name> {}` however we could maintain this and simply ignore it.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #2329 

## Summary\*

We currently allow an unlimited number of contracts in a single package despite it being generally agreed that packages should contain a single contract.

We should start enforcing this before people start relying on this behaviour. It's also much more convenient for us to encode this expectation into the type system so we don't need to pass around `Vec<(CompiledContract, DebugArtifact)>` everywhere and instead have it match binary packages.

I've then added an error which is returned if we find more than one contract within a package. Attempting to compile a contract package with multiple contracts will now print an error as below:
```
$ nargo compile
error: Packages are limited to a single contract

Error: Aborting due to 1 previous error

Location:
    tooling/nargo_cli/src/cli/mod.rs:90:5
```

## Documentation

- [x] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [x] I will request for and support Dev Rel's help in documenting this PR.

Contracts packages must only contain a single contract.

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by kevaundray** [comment]
Does this break anything in aztec-packages?

**Comment by TomAFrench** [comment]
Nope, they've been following this by convention already.

File: tooling/nargo_cli/src/cli/compile_cmd.rs [hunk 1]
```diff
@@ -227,52 +222,45 @@ fn save_program(
...
     package: &Package,
     circuit_dir: &Path,
     output_debug: bool,
 ) {
+    let (contract, debug_artifact) = contract;
+
     // TODO(#1389): I wonder if it is incorrect for nargo-core to know anything about contracts.
     // As can be seen here, It seems like a leaky abstraction where ContractFunctions (essentially CompiledPrograms)
     // are compiled via nargo-core and then the PreprocessedContract is constructed here.
     // This is due to EACH function needing it's own CRS, PKey, and VKey from the backend.
```
Comment (kevaundray) [inline tooling/nargo_cli/src/cli/compile_cmd.rs:236]: orthogonal: oof this seems outdated

Comment (TomAFrench) [reply tooling/nargo_cli/src/cli/compile_cmd.rs:236]: Just waiting on https://github.com/noir-lang/noir/pull/2666

**Review by kevaundray — DISMISSED** [review]
LGTM - I can see us needing a type alias for (CompiledContract, DebugArtifact) in the future

**Comment by TomAFrench** [comment]
> I can see us needing a type alias for (CompiledContract, DebugArtifact) in the future

Not anymore.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**